File: SkyEditor.Core/Extensions/InstalledExtensionCollection.vb
Imports System.Threading.Tasks
Imports SkyEditor
Imports SkyEditor.Core.Extensions

Namespace Extensions
    Public Class InstalledExtensionCollection
        Implements IExtensionCollection

        Public Function GetName() As Task(Of String) Implements IExtensionCollection.GetName
            Return Task.FromResult("Installed Extensions")
        End Function

        Public Function GetChildCollections(manager As PluginManager) As Task(Of IEnumerable(Of IExtensionCollection)) Implements IExtensionCollection.GetChildCollections
            Dim out As IEnumerable(Of IExtensionCollection) = manager.GetRegisteredObjects(Of ExtensionType)
            For Each item As ExtensionType In out
                item.CurrentPluginManager = manager
            Next
            Return Task.FromResult(out)
        End Function

        Public Function GetExtensionCount(manager As PluginManager) As Task(Of Integer) Implements IExtensionCollection.GetExtensionCount
            Return Task.FromResult(0)
        End Function

        Public Function GetExtensions(skip As Integer, take As Integer, manager As PluginManager) As Task(Of IEnumerable(Of ExtensionInfo)) Implements IExtensionCollection.GetExtensions
            Dim info As ExtensionInfo() = {}
            Return Task.FromResult(info.AsEnumerable)
        End Function

        Public Function InstallExtension(extensionID As String, version As String, manager As PluginManager) As Task(Of ExtensionInstallResult) Implements IExtensionCollection.InstallExtension
            Throw New NotSupportedException
        End Function

        Public Function UninstallExtension(extensionID As String, manager As PluginManager) As Task(Of ExtensionUninstallResult) Implements IExtensionCollection.UninstallExtension
            Throw New NotSupportedException
        End Function
    End Class

End Namespace

File: SkyEditor.Core/Extensions/InstalledExtensionCollection.cs
using System.Threading.Tasks;
using SkyEditor;
using SkyEditor.Core.Extensions;
using System.Collections.Generic;
using SkyEditor.Core;
using System.Linq;
using System;

namespace SkyEditor.Core.Extensions
{
    public class InstalledExtensionCollection : IExtensionCollection
    {

        public Task<string> GetName()
        {
            return Task.FromResult(Properties.Resources.UI_InstalledExtension);
        }

        public Task<IEnumerable<IExtensionCollection>> GetChildCollections(PluginManager manager)
        {
            // manager.GetRegisteredObjects returns shared objects, and it is generally not recommended to return these.
            // Creating new instances takes time, LocalExtensionCollection doesn't store any important stateful information, and there really only needs to be one of each type at any one time
            // Therefore, shared instances are OK.
            return Task.FromResult(manager.GetRegisteredObjects<LocalExtensionCollection>().AsEnumerable<IExtensionCollection>());
        }

        public Task<int> GetExtensionCount(PluginManager manager)
        {
            return Task.FromResult(0);
        }

        public Task<IEnumerable<ExtensionInfo>> GetExtensions(int skip, int take, PluginManager manager)
        {
            return Task.FromResult(Enumerable.Empty<ExtensionInfo>());
        }

        public Task<ExtensionInstallResult> InstallExtension(string extensionID, string version, PluginManager manager)
        {
            throw (new NotSupportedException());
        }

        public Task<ExtensionUninstallResult> UninstallExtension(string extensionID, PluginManager manager)
        {
            throw (new NotSupportedException());
        }
    }

}